alf/algorithms/algorithm.py

```diff
@@ -976,16 +986,29 @@ def update_with_gradient(self, loss_info, valid_masks=None, weight=1.0):
                 (``shape=(T, B), dtype=tf.float32``)
             weight (float): weight for this batch. Loss will be multiplied with
                 this weight before calculating gradient.
-
+            batch_info (BatchInfo): information about this batch returned by
+                ``ReplayBuffer.get_batch()``
         Returns:
             tuple:
             - loss_info (LossInfo): loss information.
             - params (list[(name, Parameter)]): list of parameters being updated.
         """
+        masks = None
+        if (batch_info is not None and batch_info.importance_weights != ()
+                and self._config.priority_replay_beta != 0):
+            masks = batch_info.importance_weights.pow(
+                -self._config.priority_replay_beta).unsqueeze(0)
```

Comment:
experience with 0 importance is effectively discarded. So nnz is N

Reply:
Imagine that we have an extreme case where 99 out 100 exps are 0. The replay buffer length N is still 100. A uniform replay would still treat the 100 values equally. So the importance weight should still use 1/100 instead of 1/1.

Reply:
We want E(1/importance_weight) to be 1.

Reply:
\sum p(i) * ((1/N) / p(i)) * TD(i) = \sum 1/N * TD(i) is the unbiased quantity (uniform replay). Because we are sampling according to p(i), then the ratio should be (1/N) / p(i). Because we have -beta, so the returned importance weights should be N/p(i)? And N is the replay buffer size.

Reply:
Things change if you have a 0.

Reply:
It's no longer biased. The reason of having 0 is that we really don't want that sample, hence should not include it into N.

Reply:
OK if having 0 means absolutely discarding that value then I agree NNZ is better.  I think it's depending on how we define the meaning of having a 0 priority. In the example I gave, using 100 instead of 1 as N would essentially slow down the training. But if 0 means completely discarding, then repeatedly train on that single example makes sense.

Reply:
Effectively, 0 means discarding, because it will never be sampled.